File: Core/Workspace/CSharp/Adapters/Allors.Workspace.Adapters/Workspace/Workspace.cs
// <copyright file="Workspace.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters
{
    using System.Collections.Generic;
    using Meta;
    using Numbers;

    public abstract class Workspace : IWorkspace
    {
        private readonly Dictionary<long, WorkspaceRecord> recordById;

        protected Workspace(DatabaseConnection database, IWorkspaceLifecycle lifecycle, INumbers numbers, WorkspaceIdGenerator workspaceIdGenerator)
        {
            this.Database = database;
            this.Lifecycle = lifecycle;
            this.WorkspaceIdGenerator = workspaceIdGenerator;
            this.Numbers = numbers;

            this.WorkspaceClassByWorkspaceId = new Dictionary<long, IClass>();
            this.WorkspaceIdsByWorkspaceClass = new Dictionary<IClass, long[]>();

            this.recordById = new Dictionary<long, WorkspaceRecord>();
        }

        IDatabaseConnection IWorkspace.DatabaseConnection => this.Database;
        public DatabaseConnection Database { get; }

        public IWorkspaceLifecycle Lifecycle { get; }

        public WorkspaceIdGenerator WorkspaceIdGenerator { get; }

        public INumbers Numbers { get; }

        public Dictionary<long, IClass> WorkspaceClassByWorkspaceId { get; }

        public Dictionary<IClass, long[]> WorkspaceIdsByWorkspaceClass { get; }

        public abstract ISession CreateSession();

        public WorkspaceRecord GetRecord(long id)
        {
            _ = this.recordById.TryGetValue(id, out var workspaceObject);
            return workspaceObject;
        }

        public void Push(long id, IClass @class, long version, Dictionary<IRelationType, object> changedRoleByRoleType)
        {
            if (!this.WorkspaceClassByWorkspaceId.ContainsKey(id))
            {
                this.WorkspaceClassByWorkspaceId.Add(id, @class);

                _ = this.WorkspaceIdsByWorkspaceClass.TryGetValue(@class, out var ids);
                _ = this.Numbers.Add(ids, id);
                this.WorkspaceIdsByWorkspaceClass[@class] = ids;
            }

            if (!this.recordById.TryGetValue(id, out var originalWorkspaceRecord))
            {
                this.recordById[id] = new WorkspaceRecord(@class, id, ++version, changedRoleByRoleType);
            }
            else
            {
                this.recordById[id] = new WorkspaceRecord(originalWorkspaceRecord, changedRoleByRoleType);
            }
        }
    }
}

File: typescript/libs/workspace/adapters/system/src/lib/workspace/Workspace.ts
import { IConfiguration, ISession, IWorkspace, IWorkspaceServices } from '@allors/workspace/domain/system';
import { Class, RelationType } from '@allors/workspace/meta/system';
import { importFrom } from '../collections/Range';
import { DatabaseConnection } from '../Database/DatabaseConnection';
import { Strategy } from '../session/Strategy';
import { WorkspaceRecord } from './WorkspaceRecord';

export abstract class Workspace implements IWorkspace {
  configuration: IConfiguration;

  workspaceClassByWorkspaceId: Map<number, Class>;

  workspaceIdsByWorkspaceClass: Map<Class, Set<number>>;

  private readonly recordById: Map<number, WorkspaceRecord>;

  constructor(public database: DatabaseConnection, public services: IWorkspaceServices) {
    this.configuration = database.configuration;
    this.workspaceClassByWorkspaceId = new Map();
    this.workspaceIdsByWorkspaceClass = new Map();

    this.recordById = new Map();
  }

  abstract createSession(): ISession;

  getRecord(id: number): WorkspaceRecord | undefined {
    return this.recordById.get(id);
  }

  push(id: number, cls: Class, version: number, changedRoleByRoleType: Map<RelationType, unknown> | undefined): void {
    this.workspaceClassByWorkspaceId.set(id, cls);
    let ids = this.workspaceIdsByWorkspaceClass.get(cls);
    if (ids == null) {
      ids = new Set();
      this.workspaceIdsByWorkspaceClass.set(cls, ids);
    }

    ids.add(id);

    const roleByRelationType = new Map();
    for (const [key, value] of changedRoleByRoleType) {
      if (value instanceof Strategy) {
        roleByRelationType.set(key, value.id);
      } else if (value instanceof Set) {
        roleByRelationType.set(key, importFrom([...(value as Set<Strategy>)].map((v) => v.id)));
      } else {
        roleByRelationType.set(key, value);
      }
    }

    const originalWorkspaceRecord = this.recordById.get(id);
    if (originalWorkspaceRecord == null) {
      this.recordById.set(id, new WorkspaceRecord(cls, id, ++version, roleByRelationType));
    } else {
      this.recordById.set(id, WorkspaceRecord.fromOriginal(originalWorkspaceRecord, roleByRelationType));
    }
  }
}
